### PR TITLE
fix: allocation errors on finished hp search [DET-7724]

### PR DESCRIPTION
**Pull request by eecsliu**
## Description
Sometimes a trial will report errored even though it has exited with code 0. This is because even if it has exited successfully, the daemon process might still need to be killed. This change make sure that we don't report an error after slaying the daemon. 

## Test Plan
Run some tensorflow or deepspeed hp search jobs with a `slots_per_trial` value greater than the slots in one node. Make sure that no trials error with exit code 0.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ioga — APPROVED** [review]

File: master/internal/task/allocation.go [hunk 1]
```diff
@@ -734,6 +740,16 @@ func (a *Allocation) allNonDaemonsExited() bool {
 	return true
 }
 
+func (a *Allocation) exitWithoutErr() bool {
```
Comment (ioga) [inline master/internal/task/allocation.go:743]:
```suggestion
func (a *Allocation) exitedWithoutErr() bool {
```

File: master/internal/task/allocation.go [hunk 2]
```diff
@@ -86,6 +88,7 @@ type (
 	AllocationExited struct {
 		// userRequestedStop is when a container unexpectedly exits with 0.
 		UserRequestedStop bool
+		DaemonsKilled     bool
```
Comment (ioga) [inline master/internal/task/allocation.go:91]: so that's unused?

File: master/internal/task/allocation.go [hunk 3]
```diff
@@ -49,6 +49,8 @@ type (
 		killedWhileRunning bool
 		// Marks that the trial exited successfully, but we killed some daemon containers.
 		killedDaemons bool
+		// Marks that we killed some daemon containers but after a nonzero exit.
```
Comment (ioga) [inline master/internal/task/allocation.go:52]: hmmm these comments look odd
e.g. `killedDaemons` doesn't mean the "trial exited successfully"

and `killedDaemonsNominal` looks like a zero exit, not a "nonzero".

Comment (ioga) [reply master/internal/task/allocation.go:52]: also maybe rename `killedDaemonsNominal` -> `killedDaemonsGracefully` or `killedDaemonsWithoutErr`? 
(I would struggle figuring out the meaning of the word "nominal" in such context)

Comment (eecsliu) [reply master/internal/task/allocation.go:52]: nonzero is a typo - should be zero. Will change to `gracefully`

File: master/internal/task/allocation.go (hunk 2 above)
Comment (eecsliu) [reply master/internal/task/allocation.go:91]: forgot to delete this - is vestigial from removing stuff from trial.

**Pull request merged**